ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala

```diff
@@ -4,54 +4,163 @@
 package com.daml.platform.server.api.validation
 
 import com.daml.ledger.api.domain.LedgerId
+import com.daml.ledger.grpc.GrpcStatuses
 import com.daml.platform.server.api.ApiException
-import io.grpc.{Status, StatusRuntimeException}
-
+import com.google.protobuf.{Any => AnyProto}
+import com.google.rpc.Status
+import com.google.rpc.ErrorInfo
+import io.grpc.Status.Code
+import io.grpc.StatusRuntimeException
+import io.grpc.protobuf.StatusProto
 import scalaz.syntax.tag._
 
 trait ErrorFactories {
 
+  import ErrorFactories._
+
+  def notFound(message: String): StatusRuntimeException =
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.NOT_FOUND.value())
+        .setMessage(message)
+        .addDetails(IndefiniteAnswerInfo)
+        .build()
+    )
+
   def ledgerIdMismatch(expected: LedgerId, received: LedgerId): StatusRuntimeException =
     grpcError(
-      Status.NOT_FOUND.withDescription(
-        s"Ledger ID '${received.unwrap}' not found. Actual Ledger ID is '${expected.unwrap}'."
-      )
+      Status
+        .newBuilder()
+        .setCode(Code.NOT_FOUND.value())
+        .setMessage(
+          s"Ledger ID '${received.unwrap}' not found. Actual Ledger ID is '${expected.unwrap}'."
+        )
+        .addDetails(IndefiniteAnswerInfo)
+        .build()
     )
 
   def missingField(fieldName: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Missing field: $fieldName"))
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.INVALID_ARGUMENT.value())
+        .setMessage(s"Missing field: $fieldName")
+        .addDetails(DefiniteAnswerInfo)
+        .build()
+    )
 
   def invalidArgument(errorMsg: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Invalid argument: $errorMsg"))
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.INVALID_ARGUMENT.value())
+        .setMessage(s"Invalid argument: $errorMsg")
+        .addDetails(DefiniteAnswerInfo)
+        .build()
+    )
 
   def invalidField(fieldName: String, message: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Invalid field $fieldName: $message"))
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.INVALID_ARGUMENT.value())
+        .setMessage(s"Invalid field $fieldName: $message")
+        .addDetails(DefiniteAnswerInfo)
+        .build()
+    )
 
   def outOfRange(description: String): StatusRuntimeException =
-    grpcError(Status.OUT_OF_RANGE.withDescription(description))
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.OUT_OF_RANGE.value())
+        .setMessage(description)
+        .addDetails(DefiniteAnswerInfo)
```

Comment:
AFAICT this should be an indefinite answer. Why do you think that we can meet the rank guarantee described in https://github.com/digital-asset/daml/pull/9567/files#diff-ed4ed30e4f593aaf9d013b39ed5272dfc0df08c548cc78f110fd4471eaa6f59cR143-R151 here?

Same question applies to `invalidField`, `invalidArgument`, and `missingField`.

Reply:
@andreaslochbihler-da , thanks for the link. TBH, I've been focusing on what may happen if the application retries based on usages of the "factories":
* `outOfRange` - is being used only by `LedgerOffsetValidator` which on the other hand is being used by `GetTransactionsRequest` and `GrpcCompletionStreamRequest`. What I did not take into account is that there are cases where the ledger end is calculated on the server-side. I assumed that if someone passes invalid offsets multiple times, the answer is always going to be the same. However, it may not be reasonable anymore to provide the `definite_answer` info here, as this doesn't regard command submission/deduplication.
* `invalidField` - is being used for `GrpcCompletionStreamRequest` and **command** validation. I believe if a field is not being able to be parsed, it will always result in a validation error if submitted multiple times without changes.
* `invalidArgument` - used in many places. I assumed that if an argument is invalid, it won't ever be valid, which may be a wrong assumption. Will expand this a bit more below.
* `missingField` - used similarly as `invalidField`. I assumed that if a field is missing, it won't magically appear in a re-submission, so the validation passes.
What I'm concerned about is that these "factories" may be used in more places and different contexts, so perhaps it would be reasonable to pass the `definiteAnswer: Option[Boolean]` as an argument to them. For example, I wouldn't pass it in case of the current usages of `outOfRange`. WDYT?

Reply:
> it would be reasonable to pass the `definiteAnswer: Option[Boolean]` as an argument to them

That definitely sounds reasonable.



>     * if submitted multiple times without changes.

The guarantees about `definite_answer` are independent of whether the application submits the same request or a different request. That's because it's not clear what path such a request has already taken and whether something has augmented the request on the way. The server filling in the ledger end (or maybe ledger begin elsewhere) is one example. So what may look unretryable at a certain level in our stack may actually be due to bad choices having been made during preprocessing. That's why we've decided to deal with retryability separately.

Reply:
Thanks a lot for the explanation, @andreaslochbihler-da, it's now much clearer to me.